### ISSUE TITLE
regex_replace seems to be unable to replace "X" with "\X"

### ISSUE THREAD

**Issue by m4rcinkowski** [ansible/ansible#10067]
##### Issue Type:

Bug Report
##### Ansible Version:

1.8.2
##### Environment:

Ubuntu 14.04.1 LTS
##### Summary:

This bug report is similar to #8672 but addresses a slightly different problem. All I want to do is replace one character in a string with the same character prepended with a backslash. For example: turn "lorem" into "\lorem" or "lorem+ipsum" into "lorem+ipsum".
##### Steps To Reproduce:

I've tried several different approaches. Example playbook file:

```
- hosts: 127.0.0.1
  connection: local
  vars:
    - test1: "{{ 'lorem' | regex_replace('(l)', '\\1') }}"
    - test2: "{{ 'lorem' | regex_replace('(l)', '\\ \\1') }}"
    - test3: "{{ 'lorem' | regex_replace('(l)', '\\ \\\\1') }}"
    - test4: "{{ 'lorem' | regex_replace('(l)', '\\\\\\1') }}"
  tasks:
    - debug: var={{ item }}
      with_items: [ test1, test2, test3, test4 ]
```
##### Expected Results:

I expect at least one of the `test` variables to be `\lorem`.
##### Actual Results:

```
PLAY [127.0.0.1] ************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [127.0.0.1]

TASK: [debug var={{ item }}] ************************************************** 
ok: [127.0.0.1] => (item=test1) => {
    "item": "test1", 
    "test1": "\u0001orem"
}
ok: [127.0.0.1] => (item=test2) => {
    "item": "test2", 
    "test2": "\\ \u0001orem"
}
ok: [127.0.0.1] => (item=test3) => {
    "item": "test3", 
    "test3": "\\ lorem"
}
ok: [127.0.0.1] => (item=test4) => {
    "item": "test4", 
    "test4": "\\\u0001orem"
}

PLAY RECAP ******************************************************************** 
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=0
```

**Comment by m4rcinkowski** [comment]
In fact, even an example straight from the documentation _does not work_:

```
  vars:
    - test0: "{{ 'ansible' | regex_replace('^a.*i(.*)$', 'a\\1') }}"
```

returns:

```
ok: [127.0.0.1] => (item=test0) => {
    "item": "test0", 
    "test0": "a\u0001"
}
```

Something's really messed up, guys.

**Comment by KimNorgaard** [comment]
Meanwhile you can use `\\g<1>` which seems to work.

**Comment by sivel** [comment]
The number of slashes is also dependent on whether YAML or `key=value` syntax is used.

iirc, with YAML it requires 4 slashes such as:

```
- debug:
    msg: "{{ 'ansible' | regex_replace('^a.*i(.*)$', 'a\\\\1') }}"
```

With `key=value` it requires 2 slashes:

```
- debug: msg="{{ 'ansible' | regex_replace('^a.*i(.*)$', 'a\\1') }}"
```

This is due to how the YAML and shlex parsers operate on the data.

**Comment by Gerst20051** [comment]
:+1:

**Comment by m4rcinkowski** [comment]
@KimNorgaard  Seems to work, thank you. Can you point to a reference for this `\\g<N>` workaround?
@sivel Good to know, thanks. Can you explain how can I force a backslash sign in regex_replace output? So "L" becomes "\L"?

My playbook is now:

```
- hosts: 127.0.0.1
  connection: local
  vars:
    - test1: "{{ 'ansible' | regex_replace('^a.*i(.*)$', 'a\\g<1>') }}"
    - test2: "{{ 'lorem' | regex_replace('(l)', '\\g<1>') }}"
    - test3: "{{ 'lorem' | regex_replace('(l)', '\\\\g<1>') }}"
    - test4: "{{ 'lorem' | regex_replace('(l)', '\\ \\g<1>') }}"
    - test5: "{{ 'lorem' | regex_replace('(l)', '\\ \\\\g<1>') }}"
    - test6: "{{ 'lorem' | regex_replace('(l)', '\\\\\\g<1>') }}"
  tasks:
    - debug: var=test{{ item }}
      with_sequence: count=6
```

The output it generates is:

```
PLAY [127.0.0.1] ************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [127.0.0.1]

TASK: [debug var=test{{ item }}] ********************************************** 
ok: [127.0.0.1] => (item=1) => {
    "item": "1", 
    "test1": "able"
}
ok: [127.0.0.1] => (item=2) => {
    "item": "2", 
    "test2": "lorem"
}
ok: [127.0.0.1] => (item=3) => {
    "item": "3", 
    "test3": "lorem"
}
ok: [127.0.0.1] => (item=4) => {
    "item": "4", 
    "test4": "\\ lorem"
}
ok: [127.0.0.1] => (item=5) => {
    "item": "5", 
    "test5": "\\ lorem"
}
ok: [127.0.0.1] => (item=6) => {
    "item": "6", 
    "test6": "\\g<1>orem"
}

PLAY RECAP ******************************************************************** 
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=0
```

**Comment by KimNorgaard** [comment]
@szoper Sure - https://docs.python.org/2/library/re.html

**Comment by KimNorgaard** [comment]
@szoper the replace-filter is just a wrapper around re.sub and re.compile

**Comment by sivel** [comment]
Having the python re module add a slash to the resultant string when that new slash touches a back reference is not going to really work, since you are effectively laying on escapes to the last slash (which is part of the back reference)

Additionally, if you want to only grab 1 character, you have to give it a pattern for the rest of the string, so that it gets replaced out.  Otherwise in your example you are replacing the 'l' in 'lorem' with the back reference of 'l' making the resultant string 'lorem'.  The correct pattern would be something like `(l).*`.

This is definitely a strange edge case that isn't going to be easy to solve.

**Comment by abadger** [comment]
This is currently working in the devel branch:

```
- hosts: localhost
  vars:
    - test4: "{{ 'lorem' | regex_replace('(l)', '\\\\1') }}"
    # One set of doubled backslashes is for the backslash character.  
    # The second set of doubled backslashes is part of the backreference.
    # Apparently python's re library is able to handle the string with doubled
    # backslashes that does get sent to it.

  tasks:
    - debug: var={{ item }}
      with_items:
        - "{{ test4 }}"
    # The display from debug will show double backslashes but that's
    # just the display's notation since backslashes have to themselves
    # be escaped..  It's actually a single backslash.
    - copy:
        dest: "/var/tmp/lorem"
        content: "{{ test4 }}"
    # cat'ing the resulting /var/tmp/lorem file will show that there's only a single backslash now.
```

There is some question as to whether 7f5080f64ab4a82648cb746990587c1aaff3f61d (which may be part of why this is working) is the right fix, though.  Hoping to look more at that tomorrow and then we will know whether it's time for some testing of whether this is fixed.

**Comment by abadger** [comment]
Decided it is the right fix.  Going to go ahead and close this as fixed in devel (will be the 2.0 release).
